### PR TITLE
Introduce NetworkOptions to Network and Consensus classes

### PR DESCRIPTION
**Pull request by quantumagi**
This PR:
- Introduces NetworkOptions to the Network and Consensus classes
- Updates GetHash to pass NetworkOptions
- Ensures that NetworkOptions are properly created within NetworkPeer.cs
- Modifies the usage of the ChainedBlock classes to always accept a passed hash
- Uses TemporaryOptions to round off/reduce the size of this PR

### CODE REVIEW

File: src/NBitcoin/ChainedBlock.cs [hunk 1]
```diff
@@ -75,15 +75,14 @@ public ChainedBlock(BlockHeader header, uint256 headerHash, ChainedBlock previou
         /// </summary>
         /// <param name="header">The header for the block.</param>
         /// <param name="height">The height of the block.</param>
-        public ChainedBlock(BlockHeader header, int height)
+        public ChainedBlock(BlockHeader header, uint256 headerHash, int height)
```
Comment (noescape00) [inline src/NBitcoin/ChainedBlock.cs]: Not sure if it's good idea to provide headerHash separately instead of calling `header.GetHash(); ` inside. 
Maybe it would be better to add a constructor overload so you can keep old and the new one? Looks like new constructor is needed only for a few cases but it forces it''s usage in all cases.

If you'll do this revert all the usages of the new constructor to use old constructor and use new one only when you need it.

File: src/NBitcoin/ChainedBlock.cs [hunk 2]
```diff
@@ -75,15 +75,14 @@ public ChainedBlock(BlockHeader header, uint256 headerHash, ChainedBlock previou
         /// </summary>
         /// <param name="header">The header for the block.</param>
         /// <param name="height">The height of the block.</param>
```
Comment (noescape00) [inline src/NBitcoin/ChainedBlock.cs]: Add `param name=`  for the `headerHash`

File: src/NBitcoin/Networks.cs [hunk 3]
```diff
@@ -388,11 +389,11 @@ private static Network InitStratisTest()
             messageStart[3] = 0x11;
             var magic = BitConverter.ToUInt32(messageStart, 0); //0x5223570; 
 
-            var genesis = Network.StratisMain.GetGenesis().Clone();
+            var genesis = Network.StratisMain.GetGenesis();
```
Comment (noescape00) [inline src/NBitcoin/Networks.cs]: Why was `.Clone();` removed here? 
And if it was removed here why was it left for `var consensus = Network.StratisMain.Consensus.Clone();` a few lines above?

File: src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/BlockSizeRule.cs [hunk 4]
```diff
@@ -29,7 +29,8 @@ public override Task RunAsync(ContextInformation context)
             Block block = context.BlockValidationContext.Block;
 
             // Size limits.
-            if ((block.Transactions.Count == 0) || (block.Transactions.Count > options.MaxBlockBaseSize) || (GetSize(block, NetworkOptions.None) > options.MaxBlockBaseSize))
+            if ((block.Transactions.Count == 0) || (block.Transactions.Count > options.MaxBlockBaseSize) || (GetSize(block,
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/BlockSizeRule.cs]:
```
if ((block.Transactions.Count == 0) || (block.Transactions.Count > options.MaxBlockBaseSize) ||
                (GetSize(block, context.Consensus.NetworkOptions & ~NetworkOptions.All) > options.MaxBlockBaseSize))
```
looks better then

```
if ((block.Transactions.Count == 0) || (block.Transactions.Count > options.MaxBlockBaseSize) || (GetSize(block, 
                context.Consensus.NetworkOptions & ~NetworkOptions.All) > options.MaxBlockBaseSize))
```

File: src/NBitcoin/Networks.cs (hunk 3 above)
Comment (quantumagi) [reply src/NBitcoin/Networks.cs]: GetGenesis already performs a Clone.

File: src/NBitcoin/ChainedBlock.cs (hunk 1 above)
Comment (quantumagi) [reply src/NBitcoin/ChainedBlock.cs]: The NetworkOptions required to compute the hash are more often available at a shallower call level.

File: src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/BlockSizeRule.cs (hunk 4 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/BlockSizeRule.cs]: Done.

File: src/NBitcoin/ChainedBlock.cs (hunk 2 above)
Comment (quantumagi) [reply src/NBitcoin/ChainedBlock.cs]: Done.

File: src/NBitcoin/Networks.cs (hunk 3 above)
Comment (noescape00) [reply src/NBitcoin/Networks.cs]: Hmm, makes sense, good notice. 

Since GetGenesis() performs a deserialization (cloning basically) I'd suggest renaming that so the name reflects that function will return a copy of object and not the original. Something like 
GetGenesisClone() or GetGenesisDuplicate().

Comment (quantumagi) [reply src/NBitcoin/Networks.cs]: That will affect many files. Maybe we can do that in a subsequent PR.

Comment (noescape00) [reply src/NBitcoin/Networks.cs]: Agreed. Better to make a separate PR just with this change.

**Comment by dangershony** [comment]
OMG is it really happening removing static flags?
ok let me review this closer.

File: src/NBitcoin/Block.cs [hunk 5]
```diff
@@ -106,6 +106,11 @@ public void ReadWrite(BitcoinStream stream)
         #endregion
 
         public uint256 GetHash()
+        {
+            return GetHash(NetworkOptions.TemporaryOptions);
+        }
+
+        public uint256 GetHash(NetworkOptions options)
```
Comment (dangershony) [inline src/NBitcoin/Block.cs]: This is a new method so I would rather force options to not be null and not use `TemporaryOptions inside this method.

File: src/NBitcoin/Block.cs [hunk 6]
```diff
@@ -162,21 +167,22 @@ public bool CheckProofOfWork()
 
         public bool CheckProofOfWork(Consensus consensus)
         {
-            consensus = consensus ?? Network.Main.Consensus;
+            NetworkOptions options = consensus?.NetworkOptions ?? NetworkOptions.TemporaryOptions;
```
Comment (dangershony) [inline src/NBitcoin/Block.cs]: Again lets not allow `consensus.NetworkOptions` to be null, if need be initialise it with temporary options

File: src/NBitcoin/Block.cs [hunk 7]
```diff
@@ -162,21 +167,22 @@ public bool CheckProofOfWork()
...
             if ((bits.CompareTo(BigInteger.Zero) <= 0) || (bits.CompareTo(Pow256) >= 0))
                 return false;
 
             // Check proof of work matches claimed amount.
-            if (Block.BlockSignature) // Note this can only be called on a POW block.
+            if (options.IsProofOfStake) // Note this can only be called on a POW block.
                 return this.GetPoWHash() <= this.Bits.ToUInt256();
 
-            return consensus.GetPoWHash(this) <= this.Bits.ToUInt256();
+            return consensus.GetPoWHash(consensus.NetworkOptions, this) <= this.Bits.ToUInt256();
```
Comment (dangershony) [inline src/NBitcoin/Block.cs]: We could play a bi ton this method by setting the consensus.GetPoWHash for stratis to be the x13hash
Then we can remove the `options.IsProofOfStake` check.
What do you think?

File: src/NBitcoin/Block.cs [hunk 8]
```diff
@@ -252,12 +258,11 @@ public Block(byte[] bytes)
             this.ReadWrite(bytes);
         }
 
-
         public void ReadWrite(BitcoinStream stream)
         {
             stream.ReadWrite(ref this.header);
             stream.ReadWrite(ref this.transactions);
-            if (Block.BlockSignature)
+            if ((stream.TransactionOptions ?? NetworkOptions.TemporaryOptions).IsProofOfStake)
```
Comment (dangershony) [inline src/NBitcoin/Block.cs]: Same here lets try to set the `stream.TransactionOptions` to default in `BitcoinStream` and not play around with checking for null then later its easier to change.

File: src/NBitcoin/Block.cs [hunk 9]
```diff
@@ -357,7 +368,7 @@ public bool Check()
         /// <returns></returns>
         public bool Check(Consensus consensus)
         {
-            if (Block.BlockSignature)
+            if ((consensus?.NetworkOptions ?? NetworkOptions.TemporaryOptions).IsProofOfStake)
```
Comment (dangershony) [inline src/NBitcoin/Block.cs]: Same comment  as previously

File: src/NBitcoin/Network.cs [hunk 10]
```diff
@@ -146,6 +146,7 @@ public class ConsensusOptions
         }
 
         public ConsensusOptions Options { get; set; }
+        public NetworkOptions NetworkOptions { get; set; } = NetworkOptions.TemporaryOptions;
```
Comment (dangershony) [inline src/NBitcoin/Network.cs]: This is probably wrong, we should do the same as all other networks and when building the network set the options.
Even add an overload the the NetworkBuilder to set Options.
If this is temporary and you will change it in the next PR that's fine.

File: src/NBitcoin/Network.cs [hunk 11]
```diff
@@ -195,7 +196,7 @@ public Consensus()
 
         public int SubsidyHalvingInterval { get; set; }
 
-        public Func<BlockHeader, uint256> GetPoWHash { get; set; } = h => h.GetHash();
+        public Func<NetworkOptions, BlockHeader, uint256> GetPoWHash { get; set; } = (n,h) => h.GetHash(n);
```
Comment (dangershony) [inline src/NBitcoin/Network.cs]: I dont like this delegate I would rather have an abstract (or an override on NetworkOptions that returns the `POWHash()` then stratis can override with its own part.
We can leave it to another PR but please create an issue for this.

File: src/NBitcoin/Networks.cs [hunk 12]
```diff
@@ -266,12 +266,13 @@ private static Network InitReg()
...
 
         private static Network InitStratisMain()
-        {            
+        {
             Block.BlockSignature = true;
             Transaction.TimeStamp = true;
 
             var consensus = new Consensus();
 
+            consensus.NetworkOptions = new NetworkOptions() { IsProofOfStake = true };
```
Comment (dangershony) [inline src/NBitcoin/Networks.cs]: Hmm now I am not sure about `NetworkOptions` being a property of `Consensus` arghhh...

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs [hunk 13]
```diff
@@ -577,8 +577,8 @@ private NetworkPeer(bool inbound, NetworkAddress peerAddress, Network network, N
 
             this.MessageProducer = new MessageProducer<IncomingMessage>();
 
-            this.preferredTransactionOptions = NetworkOptions.All;
-            this.SupportedTransactionOptions = NetworkOptions.None;
+            this.preferredTransactionOptions = network.NetworkOptions;
+            this.SupportedTransactionOptions = network.NetworkOptions & ~NetworkOptions.All;
```
Comment (dangershony) [inline src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs]: Don't we have a `network.NetworkOptions.All` ?

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs [hunk 14]
```diff
@@ -471,7 +471,8 @@ public void ValidateBlock(ContextInformation context)
                 // Build the next block in the chain of headers. The chain header is most likely already created by
                 // one of the peers so after we create a new chained block (mainly for validation)
                 // we ask the chain headers for its version (also to prevent memory leaks).
-                context.BlockValidationContext.ChainedBlock = new ChainedBlock(context.BlockValidationContext.Block.Header, context.BlockValidationContext.Block.Header.GetHash(), this.Tip);
+                context.BlockValidationContext.ChainedBlock = new ChainedBlock(context.BlockValidationContext.Block.Header, 
+                    context.BlockValidationContext.Block.Header.GetHash(), this.Tip);
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs]: You forgot to pass `consensus.NetworkOptions` to the `new ChainedBlock(` in GetHash()

File: src/NBitcoin/ConcurrentChain.cs [hunk 15]
```diff
@@ -24,17 +24,17 @@ public ConcurrentChain()
         {
         }
 
-        public ConcurrentChain(BlockHeader genesis)
+        public ConcurrentChain(BlockHeader genesisHeader)
         {
-            this.SetTip(new ChainedBlock(genesis, 0));
+            this.SetTip(new ChainedBlock(genesisHeader, genesisHeader.GetHash(), 0));
```
Comment (dangershony) [inline src/NBitcoin/ConcurrentChain.cs]: I am not sure you can just change the overload they do different things.

File: src/NBitcoin/ConcurrentChain.cs [hunk 16]
```diff
@@ -24,17 +24,17 @@ public ConcurrentChain()
...
         }
 
         public ConcurrentChain(Network network)
         {
             if (network != null)
             {
                 Block genesis = network.GetGenesis();
-                this.SetTip(new ChainedBlock(genesis.Header, 0));
+                this.SetTip(new ChainedBlock(genesis.Header, genesis.Header.GetHash(), 0));
```
Comment (dangershony) [inline src/NBitcoin/ConcurrentChain.cs]: Same goes here

File: src/NBitcoin/ConcurrentChain.cs [hunk 17]
```diff
@@ -71,7 +71,7 @@ public void Load(BitcoinStream stream)
                             this.blocksByHeight.Clear();
                             this.blocksById.Clear();
                             this.tip = null;
-                            this.SetTipLocked(new ChainedBlock(header, 0));
+                            this.SetTipLocked(new ChainedBlock(header, header.GetHash(), 0));
```
Comment (dangershony) [inline src/NBitcoin/ConcurrentChain.cs]: Maybe we should pass options to this constructor

File: src/NBitcoin/Block.cs (hunk 5 above)
Comment (quantumagi) [reply src/NBitcoin/Block.cs]: Done

File: src/NBitcoin/Block.cs (hunk 6 above)
Comment (quantumagi) [reply src/NBitcoin/Block.cs]: Here we are recognizing the fact that consensus itself may be null as per the legacy behavior of this method. The NetworkOptions already defaults to TemporaryOptions on the Consensus object.

File: src/NBitcoin/Block.cs (hunk 8 above)
Comment (quantumagi) [reply src/NBitcoin/Block.cs]: It's already being defaulted on the stream object. Will remove this check.

File: src/NBitcoin/Block.cs (hunk 9 above)
Comment (quantumagi) [reply src/NBitcoin/Block.cs]: Here we are recognizing the fact that consensus itself may be null as per the legacy behavior of this method. The NetworkOptions already defaults to TemporaryOptions on the Consensus object.

File: src/NBitcoin/ConcurrentChain.cs (hunk 15 above)
Comment (quantumagi) [reply src/NBitcoin/ConcurrentChain.cs]: The effect should be identical as "headerHash" replaces the "header.GetHash()" in the original method. Here the GetHash is more accessible to shallower call levels for passing in the network options - which still has to happen. Elsewhere, where the NetworkOptions are available, we are passing them in to similar calls as this one.

File: src/NBitcoin/ConcurrentChain.cs (hunk 16 above)
Comment (quantumagi) [reply src/NBitcoin/ConcurrentChain.cs]: See above. Added TODO's.

File: src/NBitcoin/Network.cs (hunk 10 above)
Comment (quantumagi) [reply src/NBitcoin/Network.cs]: It is temporary.

File: src/NBitcoin/Network.cs (hunk 11 above)
Comment (quantumagi) [reply src/NBitcoin/Network.cs]: https://github.com/stratisproject/StratisBitcoinFullNode/issues/888

File: src/Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs (hunk 14 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs]: Fixed to pass this.connectionManager.Network.(Consesnsus.)NetworkOptions.

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs (hunk 13 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs]: The constant does not include the POS flag. This should have the same effect except we include the POS flag from the network.NetworkOptions.

File: src/NBitcoin/Block.cs (hunk 7 above)
Comment (quantumagi) [reply src/NBitcoin/Block.cs]: Good idea. Made the change.

File: src/NBitcoin/ConcurrentChain.cs (hunk 17 above)
Comment (quantumagi) [reply src/NBitcoin/ConcurrentChain.cs]: This should be fine for now. Let me know if we can do that later.

File: src/NBitcoin/ConcurrentChain.cs (hunk 15 above)
Comment (dangershony) [reply src/NBitcoin/ConcurrentChain.cs]: Ok if behaviour is the same then no problem

File: src/NBitcoin/ChainBase.cs [hunk 18]
```diff
@@ -104,7 +104,7 @@ public bool TrySetTip(BlockHeader header, out ChainedBlock chainedHeader)
             if (prev == null)
                 return false;
 
-            chainedHeader = new ChainedBlock(header, header.GetHash(), this.GetBlock(header.HashPrevBlock));
+            chainedHeader = new ChainedBlock(header, header.GetHash(NetworkOptions.TemporaryOptions), this.GetBlock(header.HashPrevBlock));
```
Comment (dangershony) [inline src/NBitcoin/ChainBase.cs]: This is going to break to node as `TrySetTip` is called from `SetTip` which is called from the ChainHeaderBehaviour and is used bot in POS and POW meaning we will be using a POW Options.

We probably need to pass Options in the method, or set it somehow in ChainedBlock.

Comment (dangershony) [reply src/NBitcoin/ChainBase.cs]: I would suggest you try to sync stratis with this code to make sure I am right.

File: src/NBitcoin/ConcurrentChain.cs [hunk 19]
```diff
@@ -26,15 +26,16 @@ public ConcurrentChain()
 
         public ConcurrentChain(BlockHeader genesisHeader)
         {
+            // TODO: Pass NetworkOptions to GetHash
```
Comment (dangershony) [inline src/NBitcoin/ConcurrentChain.cs]: Ah perhaps we can pass Network in the constructor of ConcurrentChain (or ChainBase) that will solve all our problems.

File: src/NBitcoin/ChainBase.cs (hunk 18 above)
Comment (dangershony) [reply src/NBitcoin/ChainBase.cs]: See my other comment if we pass Network to ChainBase constructor then we are all good.

File: src/NBitcoin/ConcurrentChain.cs [hunk 20]
```diff
@@ -26,15 +26,16 @@ public ConcurrentChain()
...
             this.SetTip(new ChainedBlock(genesisHeader, genesisHeader.GetHash(), 0));
         }
 
         public ConcurrentChain(Network network)
         {
             if (network != null)
             {
-                Block genesis = network.GetGenesis();
-                this.SetTip(new ChainedBlock(genesis.Header, genesis.Header.GetHash(), 0));
+                Block genesis = network.GetGenesis();
```
Comment (dangershony) [inline src/NBitcoin/ConcurrentChain.cs]: Ah I see thats already the case so we can create a local Network variable

File: src/NBitcoin/ConcurrentChain.cs (hunk 19 above)
Comment (quantumagi) [reply src/NBitcoin/ConcurrentChain.cs]: Added.

File: src/NBitcoin/ConcurrentChain.cs (hunk 20 above)
Comment (quantumagi) [reply src/NBitcoin/ConcurrentChain.cs]: Added.

**Comment by quantumagi** [comment]
Don't merge yet. Tests in progress.

File: src/NBitcoin/ConcurrentChain.cs [hunk 21]
```diff
@@ -16,30 +16,29 @@ public class ConcurrentChain : ChainBase
...
+        public override Network Network { get { return this.network; } }
+        
         public ConcurrentChain()
         {
+            this.network = Network.Main;
         }
-
-        public ConcurrentChain(BlockHeader genesisHeader)
+        
+        public ConcurrentChain(BlockHeader genesisHeader, Network network = null)
```
Comment (dangershony) [inline src/NBitcoin/ConcurrentChain.cs]: Ok this has many places where its called but we have to be very careful here not to pass the wrong Network.
I would even say add a todo  to remove the `network = null` choice .

File: src/NBitcoin/ConcurrentChain.cs [hunk 22]
```diff
@@ -16,30 +16,29 @@ public class ConcurrentChain : ChainBase
...
         {
-            if (network != null)
-            {
-                Block genesis = network.GetGenesis();                
-                this.SetTip(new ChainedBlock(genesis.Header, genesis.Header.GetHash(network.NetworkOptions), 0));
-            }
         }
 
-        public ConcurrentChain(byte[] bytes)
+        public ConcurrentChain(byte[] bytes, Network network = null)
```
Comment (dangershony) [inline src/NBitcoin/ConcurrentChain.cs]: Same here.

**Comment by dangershony** [comment]
You can commit once you are sure node is stable (i.e. sync the stratis and bitcoin networks for a little while)

**Comment by quantumagi** [comment]
Ready to commit.

File: src/NBitcoin/ConcurrentChain.cs (hunk 21 above)
Comment (quantumagi) [reply src/NBitcoin/ConcurrentChain.cs]: Done

File: src/NBitcoin/ConcurrentChain.cs (hunk 22 above)
Comment (quantumagi) [reply src/NBitcoin/ConcurrentChain.cs]: Done
